make rank two and four tensors faster

rwcarlsen opened the pull request:
Store values in a single array - rather than nested arrays.  Use smarter indexing to avoid redundant arithmetic ops.  Use inlining for a frequently accessed, trivial functions. The only change that is a bit icky is the friending of RankTwo and RankFour tensors with each other.  This was necessary in order to avoid many redundant indexing arithmetic ops and I think is worth it here.

This speeds up tensor mechanics module tests by 20-30% depending on your box/environment.  Various tensor-heavy simulations run up to 15% faster (of the ones I looked at).

Fixes #9029

rwcarlsen committed:
make rank two and four tensors faster

Store values in a single array - rather than nested arrays.  Use smarter
indexing to avoid redundant arithmetic ops.

fixes #9029

permcody commented:
I'd really like to see some timing so we can have a discussion. Also, this PR will need some documentation on what's going on with some of these tweaks.

dschwen commented on modules/contact/src/ReferenceResidualProblem.C:
```diff
@@ -60,12 +60,12 @@ ReferenceResidualProblem::initialSetup()
-               ") != number of variables in system (",
-               s.n_vars(),
-               ")");
+  // if (_solnVarNames.size() > 0 && _solnVarNames.size() != s.n_vars())
```
Why is this change in here?

dschwen commented on modules/tensor_mechanics/src/utils/ElasticityTensorR4.C:
```diff
@@ -59,13 +63,19 @@ ElasticityTensorR4::elasticJacobianwc(const unsigned int i,
       for (unsigned int n = 0; n < N; ++n)
-        the_sum += _vals[i][j][m][n] * PermutationTensor::eps(m, n, k) * grad_test(j);
+        the_sum += _vals[ij1 + m * N + n] * PermutationTensor::eps(m, n, k) * grad_test(j);
+    ij1 += N * N;
```
This makes the code quite a bit harder to read. I second Cody's request for actual timings.

dschwen commented:
There is also the new `RankThreeTensor` which should not get out of sync design wise.

rwcarlsen commented:
Oops - that was for running a special sim I was benchmarking.  Will revert.

On Mon, May 1, 2017 at 10:41 AM, Daniel Schwen <notifications@github.com>
wrote:

> *@dschwen* commented on this pull request.
> ------------------------------
>
> In modules/contact/src/ReferenceResidualProblem.C
> <https://github.com/idaholab/moose/pull/9030#discussion_r114152164>:
>
> > @@ -60,12 +60,12 @@ ReferenceResidualProblem::initialSetup()
>    System & s = nonlinear_sys.system();
>    TransientExplicitSystem & as = aux_sys.sys();
>
> -  if (_solnVarNames.size() > 0 && _solnVarNames.size() != s.n_vars())
> -    mooseError("In ReferenceResidualProblem, size of solution_variables (",
> -               _solnVarNames.size(),
> -               ") != number of variables in system (",
> -               s.n_vars(),
> -               ")");
> +  // if (_solnVarNames.size() > 0 && _solnVarNames.size() != s.n_vars())
>
> Why is this change in here?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/9030#pullrequestreview-35595439>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABSeSqS_MW9HSSd7QBmLvLfcGIUT3FFsks5r1gtCgaJpZM4NNLOg>
> .
>

rwcarlsen commented:
Before:

```
...
Ran 428 tests in 124.0 seconds
428 passed, 20 skipped, 0 pending, 0 failed
```

After:

```
...
smeared_cracking.single_element............................................................. FAILED (EXODIFF)
-------------------------------------------------------------------------------------------------------------
Ran 428 tests in 99.8 seconds
427 passed, 20 skipped, 0 pending, 1 FAILED
```

That failed test runs for 1200 time steps and "passes" for the first 1000 and then begins to diverge slightly numerically - due to subgle floating point error accumulated from changed order of a few arithmetic ops.  Looks benign.  And why is there a 1200 time step test?

dschwen commented:
This is quite a substantial speedup. Nice!

rwcarlsen commented:
tensor_mechanics/examples/bridge.i (~15% speedup):

```
real	0m52.775s             real	0m44.047s
user	0m51.988s    ---->    user	0m42.825s
sys	0m0.394s              sys	0m0.791s
```

and bridge_heavy_strain.i (~10% speedup):

```
real	1m5.284s             real	0m59.150s
user	1m4.048s    ---->    user	0m57.302s
sys	0m0.743s             sys	0m1.256s
```

and an undisclosed BISON simulation (~6% speedup):

```
real	5m0.186s             real	4m42.167s
user	16m16.871s    ---->    user	16m38.646s
sys	28m32.613s             sys	28m21.181s
```

These are on my relatively modern mac+CPU.  The differences will likely be larger on older CPUs (the BISON sim, for example ran about 10% faster on a slightly older linux box).

mangerij commented:
This might conflict with my PR for migrating the PermutationTensor member
functions to the tensor classes

Or do you merge this and then I'll rework my PR?


On May 1, 2017 1:45 PM, "Robert Carlsen" <notifications@github.com> wrote:

> tensor_mechanics/examples/bridge.i (~15% speedup):
>
> real	0m52.775s             real	0m44.047s
> user	0m51.988s    ---->    user	0m42.825s
> sys	0m0.394s              sys	0m0.791s
>
> and bridge_heavy_strain.i (~10% speedup):
>
> real	1m5.284s             real	0m59.150s
> user	1m4.048s    ---->    user	0m57.302s
> sys	0m0.743s             sys	0m1.256s
>
> and an undisclosed BISON simulation (~6% speedup):
>
> real	5m0.186s             real	4m42.167s
> user	16m16.871s    ---->    user	16m38.646s
> sys	28m32.613s             sys	28m21.181s
>
> These are on my relatively modern mac+CPU. The differences will likely be
> larger on older CPUs (the BISON sim, for example ran about 10% faster on a
> slightly older linux box).
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/9030#issuecomment-298384680>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHPulCLQKcx_RVW5t_1Yh_CnPhyFulOzks5r1hosgaJpZM4NNLOg>
> .
>

rwcarlsen commented:
@mangerij - Looks like maybe a conflict on only one line.  Whichever is ready first can go in first - shouldn't be a problem.

moosebuild commented:
Job `Precheck:linux-gnu` on 3db8d3c wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 96013fe0b11729e798ca447d0799eaf5e37c00c8`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9030/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9030/style.patch | git apply -v`

rwcarlsen commented:
Okay.  This should be ready.  Do we want to regold the tensor_mechanics smeared cracking test - seems to be a small floating point error due to changes in looping/arithmetic-ops.

@bwspenc - can you take a look at the grizzly failure? - regold or is there a real problem?

dschwen commented on framework/include/utils/RankFourTensor.h:
```diff
@@ -84,13 +84,19 @@ class RankFourTensor
 
   /// Gets the value for the index specified.  Takes index = 0,1,2
-  Real & operator()(unsigned int i, unsigned int j, unsigned int k, unsigned int l);
+  inline Real & operator()(unsigned int i, unsigned int j, unsigned int k, unsigned int l)
```
The `inline` keyword is not needed when you define the function in the class body

rwcarlsen replied:
I know, but it can't hurt.  I don't really care much either way.

dschwen commented on framework/src/utils/RankThreeTensor.C:
```diff
@@ -23,6 +23,11 @@
+#define r3i(i, j, k) ((i * LIBMESH_DIM + j) * LIBMESH_DIM + k)
+#define N LIBMESH_DIM
+#define N2 (LIBMESH_DIM * LIBMESH_DIM)
+#define N3 (LIBMESH_DIM * LIBMESH_DIM * LIBMESH_DIM)
```
Why are we regressing to defines here?

rwcarlsen replied:
They are only in the C file.  It makes sure we don't waste time allocating variables, loading values from other memory locations, or doing extra math at runtime.

dschwen replied:
Sure, but `constexpr` and inline functions will behave the same way in modern compilers.

rwcarlsen replied:
removed/fixed.

dschwen commented on framework/src/utils/RankThreeTensor.C:
```diff
@@ -149,34 +136,27 @@ RankThreeTensor
-      for (unsigned int k = 0; k < N; ++k)
-        _vals[i][j][k] += a(i, j, k);
+  for (unsigned int i = 0; i < N3; ++i)
+    _vals[i] += a._vals[i];
```
How about using `std::copy` here?

rwcarlsen replied:
Good idea.

permcody replied:
Maybe... While I'm always a huge fan of using STL algorithms over hand-coded loops. This is an optimization patch. std::copy might be slower than this hand coded loop...

rwcarlsen replied:
Well - not here, but maybe for the assignment operator we could do a memcpy.

dschwen replied:
Argh, not `memcpy`. This is not 1997! And @permcody, std::copy could be faster for all you know!

dschwen replied:
http://stackoverflow.com/questions/4707012/is-it-better-to-use-stdmemcpy-or-stdcopy-in-terms-to-performance

rwcarlsen replied:
After playing with this, it doesn't seem to make much difference.  Someone else can tinker at a later time if they want.

dschwen replied:
Wait, why did I suggest `std::copy`?! This is the `+=` operator. I misread that as assignment... 😊

dschwen commented on framework/src/utils/RankThreeTensor.C:
```diff
@@ -185,23 +165,18 @@ RankThreeTensor
-      for (unsigned int k = 0; k < N; ++k)
-        result(i, j, k) = a(i, j, k) + b(i, j, k);
+  for (unsigned int i = 0; i < N3; ++i)
+    result._vals[i] = _vals[i] + b._vals[i];
```
Out of curiosity, have you checked if these loops get vectorized?

rwcarlsen replied:
On linux g++ we get no vectorization - mostly due to (potential) pointer aliasing that the compiler can't prove doesn't occur.  Not sure on mac.

dschwen commented on framework/src/utils/RankThreeTensor.C:
```diff
@@ -238,12 +207,12 @@ Real
+  for (unsigned int i = 0; i < N3; ++i)
+  {
+    Real val = _vals[i];
+    l2 += val * val;
```
I'm not at all convinced that this is any better than 

```
l2 += Utility::pow<2>(_vals[i])
```

rwcarlsen replied:
Agree - will fix.

rwcarlsen replied:
done

dschwen commented on framework/src/utils/RankThreeTensor.C:
```diff
@@ -266,31 +235,44 @@ RankThreeTensor::fillFromInputVector(const std::vector<Real> & input, FillMethod
+  unsigned int index = 0;
   for (unsigned int i = 0; i < N; ++i)
+  {
+    Real a = input(i);
```
at least make those Reals `const`

rwcarlsen replied:
done

dschwen commented on modules/tensor_mechanics/src/utils/ElasticityTensorR4.C:
```diff
@@ -59,13 +63,19 @@ ElasticityTensorR4::elasticJacobianwc(const unsigned int i,
   return the_sum * phi;
 }
 
+#define r4i(i, j, k, l) ((i * LIBMESH_DIM + j) * LIBMESH_DIM + k) * LIBMESH_DIM + l
```
👎  for the defines. This is what inline functions are for in modern C++

rwcarlsen replied:
removed.

dschwen commented on modules/tensor_mechanics/src/utils/ElasticityTensorR4.C:
```diff
@@ -78,7 +88,7 @@ ElasticityTensorR4::momentJacobian(const unsigned int i,
     for (unsigned int m = 0; m < N; ++m)
       for (unsigned int n = 0; n < N; ++n)
-        the_sum += PermutationTensor::eps(i, j, m) * _vals[j][m][k][n] * grad_phi(n);
+        the_sum += PermutationTensor::eps(i, j, m) * _vals[r4i(j, m, k, n)] * grad_phi(n);
```
this should actually just be `(*this)(j,m,k,n)` (which already _is_ inline).

rwcarlsen replied:
agree

dschwen commented on modules/tensor_mechanics/src/utils/ElasticityTensorTools.C:
```diff
@@ -21,11 +21,42 @@ elasticJacobian(const RankFourTensor & r4t,
+  Real gt2 = grad_test(2);
+  Real gp0 = grad_phi(0);
+  Real gp1 = grad_phi(1);
+  Real gp2 = grad_phi(2);
```
again, all those should be `const`. If only for the reader to know that you are basically defining shorr-hands.

rwcarlsen replied:
done

rwcarlsen commented:
@dschwen's comments should be addressed.

bwspenc commented:
@rwcarlsen I have a Grizzly patch being tested right now.

bwspenc commented:
Aaaah, gotta love smeared cracking!  Just increase the floor a little on that test.

bwspenc commented:
@rwcarlsen The patch for Grizzly is merged now, so that test failure should go away.

rwcarlsen commented:
@bwspenc - Could you make a separate PR for smeared cracking? - I'm not familiar enough with the test to know what an appropriate adjustment would be.  And does it really need to be 1200 time steps?

rwcarlsen committed:
tensor_mechanics,combined: fix broken tests

Modify smeared_cracking test to treat small numbers as zero. Skip
contact_verification/patch/ring_3.frictionless_kin as it needs the newer
version in PETSc not on all build boxes yet. and Another skipped test
too :-(

moosebuild commented:
Job `App tests:linux-gnu` on 29c4089 : invalidated by @bwspenc

Incompatible API issue has now been resolved

bwspenc commented:
@rwcarlsen Sorry, looks like you beat me to the smeared cracking fix.  The BISON test failure is due to an incompatible API issue that has now been resolved, so I invalidated the test.

It looks like you're also hitting the dreaded "Unable to located nearest node" error that occurs because the SuperLU that ships with PETSc 3.7.4 is broken (and half of our test boxes run that version of PETSc).  Unfortunately, I think you need to just skip this test along with all of the other ones we have had to skip because of this.  Add this line in the tests file for that test:
`    skip = 'Reevaluate under PETSc 3.7.4. #8200.'`
There's already another one just like that that is skipped in that tests file.

bwspenc commented:
Looks like we needed to fix MAMMOTH too.  I just submitted a PR for that.  It also looks like another contact test is broken in modules.
